libfreerdp/core/gateway/rdg.c

```diff
@@ -298,21 +305,113 @@ static BOOL rdg_write_packet(rdpRdg* rdg, wStream* sPacket)
 	return TRUE;
 }
 
-static BOOL rdg_read_all(rdpTls* tls, BYTE* buffer, int size)
+static int rdg_socket_read(BIO* bio, BYTE* pBuffer, size_t size, UINT16 transferEncoding,
+                           size_t* transferEncodingRemainingCount)
+{
+	if (transferEncoding == HTTP_TRANSFER_ENCODING_IDENTITY)
+	{
+		return BIO_read(bio, pBuffer, size);
+	}
+	else if (transferEncoding == HTTP_TRANSFER_ENCODING_CHUNKED)
+	{
+		int status;
+
+		if (transferEncodingRemainingCount == NULL)
+			return -1;
+
+		if (*transferEncodingRemainingCount == 2)
+		{
+			/* footer in buffer. get it now */
+			char _dummy[2];
+			status = BIO_read(bio, _dummy, 2);
+			if (status <= 0)
+			{
+				return status;
+			}
+			if (*transferEncodingRemainingCount >= (size_t)status)
+				*transferEncodingRemainingCount -= (size_t)status;
+			else
+				*transferEncodingRemainingCount = 0;
+		}
+
+		if (*transferEncodingRemainingCount == 0)
+		{
+			char chunkSize[11]; /* ffffffff\r\n\0 should be big enough, else fail. */
+			int _size = 0;
+			BOOL _haveNewLine = FALSE;
+			chunkSize[0] = '\0';
+
+			while (_size < 10 && !_haveNewLine)
+			{
+				status = BIO_read(bio, chunkSize + _size, 1);
+				if (status >= 0)
+				{
+					if (chunkSize[_size] == '\n')
+						_haveNewLine = TRUE;
+					_size += status;
+				}
+				else
+				{
+					if (!BIO_should_retry(bio))
+						return -1;
+					BIO_wait_read(bio, 5);
+				}
+			}
+			chunkSize[_size] = '\0';
+
+			if (status <= 0)
+			{
+				return status;
+			}
+			else
+			{
+				unsigned long tmp;
+
+				/* strtoul is tricky, error are reported via errno, we also need
+				 * to ensure the result does not overflow */
+				errno = 0;
+				tmp = strtoul(chunkSize, NULL, 16);
+				if ((errno != 0) || (tmp > SIZE_MAX - 2))
+					return -1;
+				*transferEncodingRemainingCount = tmp + 2; /* 2 bytes footer */
+				if (*transferEncodingRemainingCount == 0)
+					return -1; /* end of stream */
+			}
+		}
+		status = BIO_read(bio, pBuffer,
+		                  (size > *transferEncodingRemainingCount - 2
+		                       ? *transferEncodingRemainingCount - 2
+		                       : size));
+
+		if (status < 0)
+		{
+			return status;
+		}
+
+		*transferEncodingRemainingCount -= status;
+		return status;
+	}
+	return -1; /* should not be reached */
+}
+
+static BOOL rdg_read_all(rdpTls* tls, BYTE* buffer, size_t size, UINT16 transferEncoding,
+                         size_t* transferEncodingRemainingCount)
 {
 	int status;
 	int readCount = 0;
 	BYTE* pBuffer = buffer;
 
 	while (readCount < size)
 	{
-		status = BIO_read(tls->bio, pBuffer, size - readCount);
+		status = rdg_socket_read(tls->bio, pBuffer, size - readCount, transferEncoding,
+		                         transferEncodingRemainingCount);
 
 		if (status <= 0)
 		{
 			if (!BIO_should_retry(tls->bio))
 				return FALSE;
 
+			Sleep(10);
```

Comment:
I think we all agree that this is a hack and that the correct way to do it is polling the underlying socket

Reply:
well, I took that from https://github.com/FreeRDP/FreeRDP/pull/6733/commits/395fc6d08bf9a18f581a5882a9f39174104b3aa5

Sure, but the problem is that read_all_all does not return to the event loop.
I am by no means a openssl api expert, not even novice, but would BIO_wait_read work here?

Reply:
yeah sure the bug was already there in the previous code.